substrate/in-instructions/pallet/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+  #[cfg_attr(feature = "std", error("coin {0:?} has an invalid batch"))]
+  InvalidBatch(Coin),
+  #[cfg_attr(feature = "std", error("coin {0:?} has a different batch (ID {1})"))]
+  DifferentBatch(Coin, u64),
+}
+
+impl IsFatalError for InherentError {
+  fn is_fatal_error(&self) -> bool {
+    match self {
+      InherentError::InvalidCall | InherentError::InvalidUpdateQuantity(..) => true,
+      InherentError::UnrecognizedBlockNumber(..) => false,
+      InherentError::InvalidBlockNumber(..) => true,
+      InherentError::UnrecognizedBatches(..) => false,
+      InherentError::InvalidBatch(..) => true,
+      // One of our nodes is definitively wrong. If it's ours (signified by it passing consensus),
+      // we should panic. If it's theirs, they should be slashed
+      // Unfortunately, we can't return fatal here to trigger a slash as fatal should only be used
+      // for undeniable, technical invalidity
+      // TODO: Code a way in which this still triggers a slash vote
+      InherentError::DifferentBatch(..) => false,
+    }
+  }
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+  use super::*;
+  use frame_support::pallet_prelude::*;
+  use frame_system::pallet_prelude::*;
+
+  #[pallet::config]
+  pub trait Config: frame_system::Config<BlockNumber = u32> {
+    type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(fn deposit_event)]
+  pub enum Event<T: Config> {
+    Batch { coin: Coin, id: u64 },
+  }
+
+  #[pallet::pallet]
+  #[pallet::generate_store(pub(crate) trait Store)]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  #[pallet::storage]
+  pub(crate) type Once<T: Config> = StorageValue<_, bool, ValueQuery>;
+  #[pallet::storage]
+  #[pallet::getter(fn block_number)]
+  pub(crate) type BlockNumbers<T: Config> =
+    StorageMap<_, Blake2_256, Coin, BlockNumber, ValueQuery>;
+  #[pallet::storage]
+  #[pallet::getter(fn executed_batches)]
+  pub(crate) type NextBatch<T: Config> = StorageMap<_, Blake2_256, Coin, u64, ValueQuery>;
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_finalize(_: BlockNumberFor<T>) {
+      Once::<T>::take();
+    }
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::call_index(0)]
+    #[pallet::weight((0, DispatchClass::Mandatory))] // TODO
+    pub fn execute(origin: OriginFor<T>, updates: Updates) -> DispatchResult {
+      ensure_none(origin)?;
+      assert!(!Once::<T>::exists());
+      Once::<T>::put(true);
+
+      for (coin, update) in updates.iter().enumerate() {
+        if let Some(update) = update {
+          let coin = Coin::from(u32::try_from(coin).unwrap());
+          BlockNumbers::<T>::insert(coin, update.block_number);
+
+          for batch in &update.batches {
+            // TODO: EXECUTE
+            Self::deposit_event(Event::Batch { coin, id: batch.id });
+            NextBatch::<T>::insert(coin, batch.id + 1);
```

Comment:
Maybe we should use unique IDs instead of incremental?

Reply:
A batch is intended to be one block. I'd have no object to the ID being the block hash (represented as `[u8; 32]`). It helps resolve the issue where BlockNumbers will tell you if you fall behind, yet not if you're eclipsed. Will change :+1: